### DISCUSSION TITLE
UI Code Overhaul Proposal

### DISCUSSION THREAD

**Discussion by danaris** [endless-sky/endless-sky#9035] [Ideas]
The Endless Sky UI code is currently very individually tailored to the specific screens that it's designed to draw. There's very little modularity or ways to introduce improvements that apply to more than a single screen. Additionally, the display and interaction propagation model—putting all the Panels in a stack and having the UI singleton iterate through them to draw and trigger player interactions—is rather different than most common modern UI frameworks—at least those I am familiar with. Those use a more hierarchical model, with individual UI objects having those within them as "children", and both drawing and interactions flow through that hierarchy.

Furthermore, because each Panel is seen as being, essentially, a pane of glass taking up the entire screen (or at least, the entire ES window), they each draw to global coordinates, with no concept of relationships between different UI elements.

One consequence of all this—and the one that most proximately led me to make this proposal—is that encapsulating any given piece of the UI in a scrollable panel requires significantly rewriting the code of that individual UI panel to work directly with the scroll mechanism; there's no way to simply take an existing panel, and put it inside a scroller. Thus, all the scrolling already in the game UI is implemented locally to the panel it's scrolling, and in at least two completely different ways that I have seen so far.

I believe that reworking this UI model to conform at least somewhat more closely to a hierarchical structure, with each UI element being positioned within its parent, and then drawing using coordinates local to itself, will allow for drastically increased flexibility to modify the UI going forward. Not only will it enable the use of scrollable panes in nearly any situation (for instance, allowing the left and right panes of the Ship Info screen to scroll independently, and thus much more information to be displayed about the ship and many more outfits and cargo to be displayed), it will mean that moving any pane would only require changing its starting coordinates, rather than having every panel's Draw method specify absolute coordinates for every UI component within it (either hardcoded or stored in the `interfaces.txt` file).

It is my hope that this can even expand the functionality and flexibility of the `interfaces.txt` file.

Individual components of this proposal that I think can be considered at least somewhat separately, in roughly decreasing order of importance:

1. Redefining individual panels as representing specific rectangular spaces on the screen, and using coordinates local to those rectangles (necessary for scrollable panels)
2. Making panels hierarchical, with at least the ability of some panel types to contain others (some version of this necessary for scrollable panels)
3. Reworking the interaction code (eg, `Click()`, `Scroll()`, etc) to be called hierarchically (some modification to interaction code necessary to support scrolling)
4. Reworking the drawing code to be called hierarchically
5. Changing the coordinates used internally to match the OpenGL coordinate system, so that coders attempting to modify any of the UI code will not have to constantly switch back and forth between two coordinate systems
 	- Either instead of that or as a stopgap measure, provide methods to convert back and forth between the two systems
6. Allowing the interfaces.txt file to redefine key equivalents for buttons, rather than having those hardcoded (necessary for future work on translations)

I would love to discuss just what the most universally-helpful way of reworking the UI to be easier to work with and more modular going forward would be.

**Comment by Zitchas** [comment]
This sounds interesting and potentially useful; although I don't really know much about how the UI code works in general. I know I've often wished (as a player) that I could have various panels open simultaneously, or move them around to see stuff underneath, etc. 

I'm not sure if we have anyone with a particular UI experience/expertise, but I'd be interested in what our coders have to say about this too. 

One thing I'll note in regards to the interface.txt file. One key utility that it has is that it is data, not code. Which means that it's contents can be changed without having to recompile the game, merely restart it. That really speeds up the testing and development process; as well as allowing players and non-coders to do things like, for instance, change the colors used for various things for themselves

**Comment by danaris** [comment]
Yeah; I'm still not at all sure just how _much_ I can feasibly expand the capabilities of the `interfaces.txt` file, but I definitely think it's worth giving a serious look to.

I think what would be particularly neat is if I could get it to support essentially taking pieces of the existing user interface and cobbling them together into new shapes—for instance, I saw another PR, #6822, that takes the ship outline with weapon hardpoint indicators from the Ship Info panel and puts it on a new panel; it would be really neat if this were something people could just stick on a new custom panel.

How that would work in implementation, and whether it would be possible to extent the datafile enough to actually be able to build meaningful navigation abilities with it without extra code...I'm not sure. My plan is to take this one step at a time, and see how far it can get us.

**Comment by danaris** [comment]
Oh; one other thing that occurred to me about this is that we can also almost certainly increase performance with some fairly straightforward changes, as right now, _everything_ in the game draws every frame. Not just the parts that have changed; every single UI element redraws every pixel every frame. 

Obviously, for the main interface itself (the player flying around), there's not going to be enough that doesn't update each frame for it to be very worth trying to optimize any of it. But for most UI panes, updates happen either never (they come on the screen in one state, and keep that state until they leave the screen), or in response to specific user inputs, which shouldn't be too hard to keep track of and push updates in response to.

**Comment by thewierdnut** [comment]
As a proof of concept, It is possible to generate simple resizable panels entirely in the shader: https://github.com/thewierdnut/endless-mobile/blob/android/source/UiRectShader.cpp
